File: slp/backend-dotnet/Features/Search/SearchService.cs
using backend_dotnet.Data;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.Search;

/// <summary>
///   Simple search using ILIKE (case‑insensitive substring match) on:
///     - Quizzes   → title
///     - Questions → content
///     - Sources   → title
///     - Favorites → text
/// </summary>
public class SearchService : ISearchService
{
    private readonly AppDbContext _context;
    private readonly ILogger<SearchService> _logger;

    private static readonly HashSet<string> ValidTypes =
        new(StringComparer.OrdinalIgnoreCase) { "all", "quiz", "question", "source", "favorite" };

    public SearchService(AppDbContext context, ILogger<SearchService> logger)
    {
        _context = context;
        _logger = logger;
    }

    // ── Public entry point ───────────────────────────────────────────────────

    public async Task<SearchResponse> SearchAsync(SearchRequest req, int userId)
    {
        var q = (req.Q ?? string.Empty).Trim();
        var page = Math.Max(1, req.Page);
        var pageSize = Math.Clamp(req.PageSize, 1, 50);
        var offset = (page - 1) * pageSize;
        var type = ValidTypes.Contains(req.Type ?? "all")
                         ? req.Type!.ToLowerInvariant()
                         : "all";

        if (q.Length == 0)
        {
            return new SearchResponse
            {
                Query = q,
                Type = type,
                Page = page,
                PageSize = pageSize,
                TotalCount = 0,
                TotalPages = 0,
                Results = new List<SearchResultItem>()
            };
        }

        return type switch
        {
            "quiz" => await SearchSingleTypeAsync(q, userId, page, pageSize, offset, "quiz"),
            "question" => await SearchSingleTypeAsync(q, userId, page, pageSize, offset, "question"),
            "source" => await SearchSingleTypeAsync(q, userId, page, pageSize, offset, "source"),
            "favorite" => await SearchSingleTypeAsync(q, userId, page, pageSize, offset, "favorite"),
            _ => await SearchAllAsync(q, userId, pageSize),
        };
    }

    // ── Single-type search (with pagination) ─────────────────────────────────

    private async Task<SearchResponse> SearchSingleTypeAsync(
        string q, int userId, int page, int pageSize, int offset, string type)
    {
        var (items, totalCount) = type switch
        {
            "quiz" => await SearchQuizzesAsync(q, userId, offset, pageSize),
            "question" => await SearchQuestionsAsync(q, userId, offset, pageSize),
            "source" => await SearchSourcesAsync(q, userId, offset, pageSize),
            "favorite" => await SearchFavoritesAsync(q, userId, offset, pageSize),
            _ => (new List<SearchResultItem>(), 0),
        };

        return new SearchResponse
        {
            Query = q,
            Type = type,
            Page = page,
            PageSize = pageSize,
            TotalCount = totalCount,
            TotalPages = totalCount == 0 ? 0 : (int)Math.Ceiling(totalCount / (double)pageSize),
            Results = items,
        };
    }

    // ── "All" search: fetch up to pageSize from each category, merge by rank ─
    //    (Rank is now a simple 1.0 for all matches – ordering is by created_at desc)

    private async Task<SearchResponse> SearchAllAsync(string q, int userId, int pageSize)
    {
        var (quizItems, quizCount) = await SearchQuizzesAsync(q, userId, 0, pageSize);
        var (questionItems, questionCount) = await SearchQuestionsAsync(q, userId, 0, pageSize);
        var (sourceItems, sourceCount) = await SearchSourcesAsync(q, userId, 0, pageSize);
        var (favoriteItems, favoriteCount) = await SearchFavoritesAsync(q, userId, 0, pageSize);

        // Merge and sort by most recent first (no rank scoring)
        var merged = quizItems
            .Concat(questionItems)
            .Concat(sourceItems)
            .Concat(favoriteItems)
            .OrderByDescending(r => r.CreatedAt)
            .Take(pageSize)
            .ToList();

        var totalCount = quizCount + questionCount + sourceCount + favoriteCount;

        return new SearchResponse
        {
            Query = q,
            Type = "all",
            Page = 1,
            PageSize = pageSize,
            TotalCount = totalCount,
            TotalPages = 1,
            Results = merged,
            CategoryCounts = new CategoryCounts
            {
                Quizzes = quizCount,
                Questions = questionCount,
                Sources = sourceCount,
                Favorites = favoriteCount,
            },
        };
    }

    // ── Per-category search methods (simple ILIKE) ───────────────────────────

    private async Task<(List<SearchResultItem>, int)> SearchQuizzesAsync(
        string q, int userId, int offset, int limit)
    {
        try
        {
            // Build the pattern for ILIKE
            var pattern = $"%{q}%";

            var query = _context.Quizzes
                .Where(quiz => !quiz.Disabled &&
                               (quiz.Visibility == "public" || quiz.UserId == userId) &&
                               EF.Functions.ILike(quiz.Title, pattern))
                .OrderByDescending(quiz => quiz.CreatedAt)
                .Skip(offset)
                .Take(limit)
                .Select(quiz => new
                {
                    quiz.Id,
                    quiz.Title,
                    quiz.Visibility,
                    quiz.CreatedAt,
                    Snippet = quiz.Title.Length > 200
                        ? quiz.Title.Substring(0, 200) + "..."
                        : quiz.Title
                });

            var itemsList = await query.ToListAsync();

            // Get total count separately (EF Core doesn't support COUNT(*) OVER() easily with Skip/Take)
            var totalCount = await _context.Quizzes
                .Where(quiz => !quiz.Disabled &&
                               (quiz.Visibility == "public" || quiz.UserId == userId) &&
                               EF.Functions.ILike(quiz.Title, pattern))
                .CountAsync();

            if (itemsList.Count == 0)
                return (new List<SearchResultItem>(), 0);

            var quizIds = itemsList.Select(x => x.Id).ToList();

            // Load tags for display
            var tagLookup = (await _context.QuizTags
                .Where(qt => quizIds.Contains(qt.QuizId))
                .Select(qt => new { qt.QuizId, qt.Tag.Name })
                .ToListAsync())
                .ToLookup(x => x.QuizId, x => x.Name);

            var items = itemsList.Select(r => new SearchResultItem
            {
                ResultType = "quiz",
                Id = r.Id,
                Title = r.Title,
                Snippet = r.Snippet,
                Rank = 1.0,          // no relevance scoring
                Visibility = r.Visibility,
                CreatedAt = r.CreatedAt,
                Tags = tagLookup[r.Id].ToList(),
            }).ToList();

            return (items, totalCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching quizzes for query '{Query}'", q);
            return (new List<SearchResultItem>(), 0);
        }
    }

    private async Task<(List<SearchResultItem>, int)> SearchQuestionsAsync(
        string q, int userId, int offset, int limit)
    {
        try
        {
            var pattern = $"%{q}%";

            var query = _context.Questions
                .Where(question => question.UserId == userId &&
                                   EF.Functions.ILike(question.Content, pattern))
                .OrderByDescending(question => question.CreatedAt)
                .Skip(offset)
                .Take(limit)
                .Select(question => new
                {
                    question.Id,
                    Title = question.Content.Length > 120
                        ? question.Content.Substring(0, 120)
                        : question.Content,
                    question.Type,
                    question.CreatedAt,
                    Snippet = question.Content.Length > 200
                        ? question.Content.Substring(0, 200) + "..."
                        : question.Content
                });

            var itemsList = await query.ToListAsync();

            var totalCount = await _context.Questions
                .Where(question => question.UserId == userId &&
                                   EF.Functions.ILike(question.Content, pattern))
                .CountAsync();

            if (itemsList.Count == 0)
                return (new List<SearchResultItem>(), 0);

            var questionIds = itemsList.Select(x => x.Id).ToList();

            var tagLookup = (await _context.QuestionTags
                .Where(qt => questionIds.Contains(qt.QuestionId))
                .Select(qt => new { qt.QuestionId, qt.Tag.Name })
                .ToListAsync())
                .ToLookup(x => x.QuestionId, x => x.Name);

            var items = itemsList.Select(r => new SearchResultItem
            {
                ResultType = "question",
                Id = r.Id,
                Title = r.Title,
                Snippet = r.Snippet,
                Rank = 1.0,
                SubType = r.Type,
                CreatedAt = r.CreatedAt,
                Tags = tagLookup[r.Id].ToList(),
            }).ToList();

            return (items, totalCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching questions for query '{Query}'", q);
            return (new List<SearchResultItem>(), 0);
        }
    }

    private async Task<(List<SearchResultItem>, int)> SearchSourcesAsync(
        string q, int userId, int offset, int limit)
    {
        try
        {
            var pattern = $"%{q}%";

            var query = _context.Sources
                .Where(source => source.UserId == userId &&
                                 source.DeletedAt == null &&
                                 EF.Functions.ILike(source.Title, pattern))
                .OrderByDescending(source => source.CreatedAt)
                .Skip(offset)
                .Take(limit)
                .Select(source => new
                {
                    source.Id,
                    source.Title,
                    source.Type,
                    source.CreatedAt,
                    Snippet = source.Title.Length > 200
                        ? source.Title.Substring(0, 200) + "..."
                        : source.Title
                });

            var itemsList = await query.ToListAsync();

            var totalCount = await _context.Sources
                .Where(source => source.UserId == userId &&
                                 source.DeletedAt == null &&
                                 EF.Functions.ILike(source.Title, pattern))
                .CountAsync();

            if (itemsList.Count == 0)
                return (new List<SearchResultItem>(), 0);

            var items = itemsList.Select(r => new SearchResultItem
            {
                ResultType = "source",
                Id = r.Id,
                Title = r.Title,
                Snippet = r.Snippet,
                Rank = 1.0,
                SubType = r.Type,
                CreatedAt = r.CreatedAt,
                Tags = new List<string>(),
            }).ToList();

            return (items, totalCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching sources for query '{Query}'", q);
            return (new List<SearchResultItem>(), 0);
        }
    }

    private async Task<(List<SearchResultItem>, int)> SearchFavoritesAsync(
        string q, int userId, int offset, int limit)
    {
        try
        {
            var pattern = $"%{q}%";

            var query = _context.FavoriteItems
                .Where(fav => fav.UserId == userId &&
                              EF.Functions.ILike(fav.Text, pattern))
                .OrderByDescending(fav => fav.CreatedAt)
                .Skip(offset)
                .Take(limit)
                .Select(fav => new
                {
                    fav.Id,
                    fav.Text,
                    fav.Type,
                    fav.CreatedAt,
                    Snippet = fav.Text.Length > 200
                        ? fav.Text.Substring(0, 200) + "..."
                        : fav.Text
                });

            var itemsList = await query.ToListAsync();

            var totalCount = await _context.FavoriteItems
                .Where(fav => fav.UserId == userId &&
                              EF.Functions.ILike(fav.Text, pattern))
                .CountAsync();

            if (itemsList.Count == 0)
                return (new List<SearchResultItem>(), 0);

            var items = itemsList.Select(r => new SearchResultItem
            {
                ResultType = "favorite",
                Id = r.Id,
                Title = r.Text,
                Snippet = r.Snippet,
                Rank = 1.0,
                SubType = r.Type,
                CreatedAt = r.CreatedAt,
                Tags = new List<string>(),
            }).ToList();

            return (items, totalCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching favorites for query '{Query}'", q);
            return (new List<SearchResultItem>(), 0);
        }
    }
}

File: slp/backend-nestjs/src/modules/search/search.service.ts
import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository, ILike, In } from "typeorm";
import { Quiz } from "../quiz/quiz.entity";
import { Question } from "../question/question.entity";
import { Source } from "../source/source.entity";
import { FavoriteItem } from "../favorite/favorite-item.entity";
import { QuizTag } from "../quiz/quiz-tag.entity";
import { QuestionTag } from "../question/question-tag.entity";

export interface SearchParams {
  q: string;
  type: string;
  page: number;
  pageSize: number;
  offset: number;
  userId: number;
}

@Injectable()
export class SearchService {
  constructor(
    @InjectRepository(Quiz)
    private quizRepo: Repository<Quiz>,
    @InjectRepository(Question)
    private questionRepo: Repository<Question>,
    @InjectRepository(Source)
    private sourceRepo: Repository<Source>,
    @InjectRepository(FavoriteItem)
    private favoriteRepo: Repository<FavoriteItem>,
    @InjectRepository(QuizTag)
    private quizTagRepo: Repository<QuizTag>,
    @InjectRepository(QuestionTag)
    private questionTagRepo: Repository<QuestionTag>,
  ) {}

  async search(params: SearchParams): Promise<any> {
    const { q, type, page, pageSize, offset, userId } = params;
    if (type !== "all") {
      return this.searchSingleType(q, userId, page, pageSize, offset, type);
    }
    return this.searchAll(q, userId, pageSize);
  }

  // -------------------------------------------------------------------------
  // Single type with proper pagination (same as .NET)
  // -------------------------------------------------------------------------
  private async searchSingleType(
    q: string,
    userId: number,
    page: number,
    pageSize: number,
    offset: number,
    type: string,
  ) {
    let items: any[] = [];
    let totalCount = 0;

    switch (type) {
      case "quiz":
        const quizRes = await this.searchQuizzes(q, userId, offset, pageSize);
        items = quizRes.items;
        totalCount = quizRes.totalCount;
        break;
      case "question":
        const questionRes = await this.searchQuestions(
          q,
          userId,
          offset,
          pageSize,
        );
        items = questionRes.items;
        totalCount = questionRes.totalCount;
        break;
      case "source":
        const sourceRes = await this.searchSources(q, userId, offset, pageSize);
        items = sourceRes.items;
        totalCount = sourceRes.totalCount;
        break;
      case "favorite":
        const favRes = await this.searchFavorites(q, userId, offset, pageSize);
        items = favRes.items;
        totalCount = favRes.totalCount;
        break;
    }

    return {
      query: q,
      type,
      page,
      pageSize,
      totalCount,
      totalPages: totalCount === 0 ? 0 : Math.ceil(totalCount / pageSize),
      results: items,
    };
  }

  // -------------------------------------------------------------------------
  // "all" mode: fetch up to pageSize from each category, merge by created_at DESC
  // -------------------------------------------------------------------------
  private async searchAll(q: string, userId: number, pageSize: number) {
    const quizzes = await this.searchQuizzes(q, userId, 0, pageSize);
    const questions = await this.searchQuestions(q, userId, 0, pageSize);
    const sources = await this.searchSources(q, userId, 0, pageSize);
    const favorites = await this.searchFavorites(q, userId, 0, pageSize);

    const merged = [
      ...quizzes.items,
      ...questions.items,
      ...sources.items,
      ...favorites.items,
    ]
      .sort(
        (a, b) =>
          new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime(),
      )
      .slice(0, pageSize);

    const totalCount =
      quizzes.totalCount +
      questions.totalCount +
      sources.totalCount +
      favorites.totalCount;

    return {
      query: q,
      type: "all",
      page: 1,
      pageSize,
      totalCount,
      totalPages: 1,
      results: merged,
      categoryCounts: {
        quizzes: quizzes.totalCount,
        questions: questions.totalCount,
        sources: sources.totalCount,
        favorites: favorites.totalCount,
      },
    };
  }

  // -------------------------------------------------------------------------
  // Quiz search – identical to .NET: public + user's own, ILIKE on title
  // -------------------------------------------------------------------------
  private async searchQuizzes(
    q: string,
    userId: number,
    offset: number,
    limit: number,
  ) {
    const pattern = `%${q}%`;
    const queryBuilder = this.quizRepo
      .createQueryBuilder("quiz")
      .where("quiz.disabled = false")
      .andWhere("(quiz.visibility = :public OR quiz.userId = :userId)", {
        public: "public",
        userId,
      })
      .andWhere("quiz.title ILIKE :pattern", { pattern })
      .orderBy("quiz.createdAt", "DESC")
      .skip(offset)
      .take(limit);

    const [quizzes, totalCount] = await queryBuilder.getManyAndCount();

    if (quizzes.length === 0) return { items: [], totalCount };

    const quizIds = quizzes.map((q) => q.id);
    const quizTags = await this.quizTagRepo.find({
      where: { quizId: In(quizIds) },
      relations: ["tag"],
    });
    const tagMap = new Map<number, string[]>();
    for (const qt of quizTags) {
      if (!tagMap.has(qt.quizId)) tagMap.set(qt.quizId, []);
      tagMap.get(qt.quizId)!.push(qt.tag.name);
    }

    const items = quizzes.map((quiz) => ({
      resultType: "quiz",
      id: quiz.id,
      title: quiz.title,
      snippet:
        quiz.title.length > 200
          ? quiz.title.substring(0, 200) + "..."
          : quiz.title,
      rank: 1.0,
      tags: tagMap.get(quiz.id) || [],
      createdAt: quiz.createdAt,
      subType: null,
      visibility: quiz.visibility,
    }));

    return { items, totalCount };
  }

  // -------------------------------------------------------------------------
  // Question search – user's own, ILIKE on content
  // -------------------------------------------------------------------------
  private async searchQuestions(
    q: string,
    userId: number,
    offset: number,
    limit: number,
  ) {
    const pattern = `%${q}%`;
    const queryBuilder = this.questionRepo
      .createQueryBuilder("question")
      .where("question.userId = :userId", { userId })
      .andWhere("question.content ILIKE :pattern", { pattern })
      .orderBy("question.createdAt", "DESC")
      .skip(offset)
      .take(limit);

    const [questions, totalCount] = await queryBuilder.getManyAndCount();

    if (questions.length === 0) return { items: [], totalCount };

    const questionIds = questions.map((q) => q.id);
    const questionTags = await this.questionTagRepo.find({
      where: { questionId: In(questionIds) },
      relations: ["tag"],
    });
    const tagMap = new Map<number, string[]>();
    for (const qt of questionTags) {
      if (!tagMap.has(qt.questionId)) tagMap.set(qt.questionId, []);
      tagMap.get(qt.questionId)!.push(qt.tag.name);
    }

    const items = questions.map((q) => ({
      resultType: "question",
      id: q.id,
      title: q.content.length > 120 ? q.content.substring(0, 120) : q.content,
      snippet:
        q.content.length > 200
          ? q.content.substring(0, 200) + "..."
          : q.content,
      rank: 1.0,
      tags: tagMap.get(q.id) || [],
      createdAt: q.createdAt,
      subType: q.type,
      visibility: null,
    }));

    return { items, totalCount };
  }

  // -------------------------------------------------------------------------
  // Source search – user's own, not deleted, ILIKE on title
  // -------------------------------------------------------------------------
  private async searchSources(
    q: string,
    userId: number,
    offset: number,
    limit: number,
  ) {
    const pattern = `%${q}%`;
    const [sources, totalCount] = await this.sourceRepo
      .createQueryBuilder("source")
      .where("source.userId = :userId", { userId })
      .andWhere("source.deletedAt IS NULL")
      .andWhere("source.title ILIKE :pattern", { pattern })
      .orderBy("source.createdAt", "DESC")
      .skip(offset)
      .take(limit)
      .getManyAndCount();

    const items = sources.map((s) => ({
      resultType: "source",
      id: s.id,
      title: s.title,
      snippet:
        s.title.length > 200 ? s.title.substring(0, 200) + "..." : s.title,
      rank: 1.0,
      tags: [],
      createdAt: s.createdAt,
      subType: s.type,
      visibility: null,
    }));

    return { items, totalCount };
  }

  // -------------------------------------------------------------------------
  // Favorite search – user's own, ILIKE on text
  // -------------------------------------------------------------------------
  private async searchFavorites(
    q: string,
    userId: number,
    offset: number,
    limit: number,
  ) {
    const pattern = `%${q}%`;
    const [favorites, totalCount] = await this.favoriteRepo
      .createQueryBuilder("fav")
      .where("fav.userId = :userId", { userId })
      .andWhere("fav.text ILIKE :pattern", { pattern })
      .orderBy("fav.createdAt", "DESC")
      .skip(offset)
      .take(limit)
      .getManyAndCount();

    const items = favorites.map((f) => ({
      resultType: "favorite",
      id: f.id,
      title: f.text,
      snippet: f.text.length > 200 ? f.text.substring(0, 200) + "..." : f.text,
      rank: 1.0,
      tags: [],
      createdAt: f.createdAt,
      subType: f.type,
      visibility: null,
    }));

    return { items, totalCount };
  }
}
